RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,644 @@
+            op.tensor = at::empty(shape_, op.options(), MemoryFormat::Contiguous);
+            op.current_dtype = op.target_dtype;
+          } else if (op.will_resize) {
+            at::native::resize_output(op.tensor, shape_);
+          }
+          */
+          // AFTER:
+          set_output(i, shape_, {}, op.options(), names_);
+        }
+        break;
+      }
+```
+
+TensorIterator defines an override of `set_output` that recovers the old
+behavior, while structured kernel subclasses override `set_output` in
+the same way as before.
+
+The code generation only requires a very modest extension: an `structured_inherits`
+field that lets you replace `MetaBase` with your own custom base
+implementation class:
+
+```
+- func: add.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True
+  structured_inherits: TensorIteratorBase  # [NEW]
+  dispatch:
+    CPU: upsample_nearest1d_structured_cpu
+    CUDA: upsample_nearest1d_structured_cuda
+```
+
+Now you can simply construct it appropriately in your function
+definitions:
+
+```
+TORCH_META_FUNC2(add, Tensor) (
+  const Tensor& self, const Tensor& other, Scalar alpha
+) {
+  build_binary_op(maybe_get_output(), self, other);
+  native::alpha_check(dtype(), alpha);
+}
+
+TORCH_IMPL_FUNC(add_out) (
+  Tensor& result, const Tensor& self, const Tensor& other, Scalar alpha
+) {
+  add_stub(device_type(), *this, alpha);
+  TORCH_INTERNAL_ASSERT(result.scalar_type() == output().dtype());
+}
+```
+
+In the case of TensorIterator, all of the arguments are stored in the
+struct at construction time, so the impl func doesn't need to make
+use of any of the tensors.  Additionally, when `set_output` is invoked
+in the meta function, after doing all appropriate allocations, it will
+delegate to the underlying `set_output` in TensorIterator, letting it
+query with `maybe_get_output()` to register any outputs necessary.
+
+## Piggybacking other improvements
+
+Ports to structured kernels must be done individually by hand.  Because
+the porting process is already labor intensive, we should do other
+improvements "while the patient is open".  Here are some candidate
+improvements which should consider applying
+
+### Removing mutable references from out arguments
+
+Historically, out and inplace variants of functions take a `Tensor&`
+rather than a `const Tensor&`.  This convention has lead to no end
+of confusion for kernel writers, who incorrectly surmise that given
+a mutable reference `Tensor& out`, one may assign the output by
+writing `out = ... some expression ...`  (this doesn't work).
+
+The absence of the const modifier is currently relied upon by
+template metaprogramming machinery (to detect if arguments are out
+tensors or not); however, because the implementations of structured
+kernels are a layer below the operator registration layer, the
+const modifier can be eliminated from the `TORCH_IMPL_FUNC` API
```

Comment:
I think you mean the const modifier can be added, not eliminated, right? Like, out arguments can now be typed as `const Tensor&` in `TORCH_IMPL_FUNC` declarations, as in
```
TORCH_IMPL_FUNC(upsample_nearest1d_out_cpu) (
    const Tensor& input,
    IntArrayRef output_size,
    c10::optional<double> scales,
    const Tensor& output             // currently just Tensor& output
) {
  upsample_nearest1d_kernel(kCPU, output, input, scales);
}
```
Aside: it happens that `upsample_nearest1d_kernel` itself takes `Tensor& output`, which might hint at additional porting work for at least some kernels to adopt this change.